plutus-ir/src/Language/PlutusIR/Optimizer/DeadCode.hs

```diff
@@ -0,0 +1,146 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+-- | Optimization passes for removing dead code, mainly dead let bindings.
+module Language.PlutusIR.Optimizer.DeadCode (removeDeadBindings) where
+
+import           Language.PlutusIR
+
+import qualified Language.PlutusCore      as PLC
+import qualified Language.PlutusCore.Name as PLC
+
+import           Control.Monad
+import           Control.Monad.State
+
+import           Data.Coerce
+import qualified Data.Map                 as Map
+
+import           Lens.Micro
+
+-- | Remove all the dead let bindings in a term.
+removeDeadBindings
+    :: (PLC.HasUnique (name a) PLC.TermUnique, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Term tyname name a
+    -> Term tyname name a
+removeDeadBindings t = evalState (processTerm t) mempty
+
+type Usages = Map.Map PLC.Unique Int
+
+addUsage :: (PLC.HasUnique n unique) => n -> Usages -> Usages
+addUsage n usages =
+    let
+        u = coerce $ n ^. PLC.unique
+        old = Map.findWithDefault 0 u usages
+    in Map.insert u (old+1) usages
+
+hasUsage :: (PLC.HasUnique n unique) => n -> Usages -> Bool
+hasUsage n usages = Map.findWithDefault 0 (coerce $ n ^. PLC.unique) usages > 0
+
+liveBinding
+    :: (MonadState Usages m, PLC.HasUnique (name a) PLC.TermUnique, PLC.HasUnique (tyname a) PLC.TypeUnique)
+    => Binding tyname name a
+    -> m Bool
+liveBinding b =
+    let
+        -- TODO: HasUnique instances for VarDecl and TyVarDecl?
+        hasUsageVarDecl (VarDecl _ n _) = hasUsage n
+        hasUsageTyVarDecl (TyVarDecl _ n _) = hasUsage n
+    in case b of
+        TermBind _ d _ -> gets $ hasUsageVarDecl d
+        TypeBind _ d _ -> gets $ hasUsageTyVarDecl d
+        DatatypeBind _ (Datatype _ d _ destr constrs) -> do
+            usages <- get
+            let used = hasUsageTyVarDecl d usages|| hasUsage destr usages || any (\c -> hasUsageVarDecl c usages) constrs
+            pure used
+
+{- Note [Simultaneous usage analysis and dead binding elimination]
+It's tempting to say that usage analysis and dead binding elmination should be
+separate passes. We could proceed for each let by:
+- Processing the body.
+- Getting the usages for the body.
+- Removing dead bindings.
+
+But this is quadratic, because it traverses subterms many times. The problem is that
+the usage information for subterms remains relevant when processing superterms, and
+it is very wasteful to recompute it. Possibly there is an elegant way to separate
```

Comment:
Can we store usage information in `Term` annotations?

Reply:
Not terribly easily - remember that the information changes as we remove dead bindings because usages go away! So if we're using this style of usage information we have to compute it in an interleaved fashion (but maybe we could somehow separate out the code).

If we wanted to compute it statically up front, then we'd need some kind of dependency graph representation which would account for transitive deadness. Maybe I should just rewrite it to take that approach, but I'd like to get something working relatively soon.

Reply:
I'm personally fine with the current design. Maybe we'll need more granularity in future, but right now this seems fine.

Reply:
I have an inkling that I could make a very elegant implementation using algebraic graphs, but I'll leave the fun for another time ;)